utils/src/ParticleIDUtils.cc

```diff
@@ -0,0 +1,139 @@
+#include <edm4hep/utils/ParticleIDUtils.h>
+
+#include "edm4hep/Constants.h"
+
+#include <podio/FrameCategories.h>
+
+#include <cassert>
+#include <iterator>
+#include <stdexcept>
+#include <string>
+
+namespace edm4hep::utils {
+
+std::optional<int> getParamIndex(const ParticleIDMeta& pidMetaInfo, const std::string& param) {
+  const auto nameIt = std::find(pidMetaInfo.paramNames.begin(), pidMetaInfo.paramNames.end(), param);
+  if (nameIt != pidMetaInfo.paramNames.end()) {
+    return std::distance(pidMetaInfo.paramNames.begin(), nameIt);
+  }
+  return std::nullopt;
+}
+
+void PIDHandler::addColl(const edm4hep::ParticleIDCollection& coll) {
+  for (const auto pid : coll) {
+    m_recoPidMap.emplace(pid.getParticle(), pid);
+  }
+}
+
+void PIDHandler::addColl(const edm4hep::ParticleIDCollection& coll, const edm4hep::utils::ParticleIDMeta& pidInfo) {
+  addColl(coll);
+  addMetaInfo(pidInfo);
+}
+
+void PIDHandler::addMetaInfo(const edm4hep::utils::ParticleIDMeta& pidInfo) {
+  const auto [algoIt, inserted] = m_algoTypes.emplace(pidInfo.algoName, pidInfo.algoType);
+  if (!inserted) {
+    throw std::runtime_error("Cannot have duplicate algorithm names (" + pidInfo.algoName + " already exists)");
+  }
+
+  const auto [__, metaInserted] = m_algoPidMeta.emplace(pidInfo.algoType, pidInfo);
+  if (!metaInserted) {
+    if (inserted) {
+      m_algoTypes.erase(algoIt);
+    }
+    throw std::runtime_error("Cannot have duplicate algorithm types (" + std::to_string(pidInfo.algoType) +
+                             " already exists)");
+  }
+}
+
+PIDHandler PIDHandler::from(const podio::Frame& event, const podio::Frame& metadata) {
+  PIDHandler handler{};
+  for (const auto& name : event.getAvailableCollections()) {
+    const auto* coll = event.get(name);
+    if (const auto pidColl = dynamic_cast<const edm4hep::ParticleIDCollection*>(coll)) {
+      handler.addColl(*pidColl);
+
+      auto maybeMetaInfo = PIDHandler::getAlgoInfo(metadata, name);
+      if (maybeMetaInfo) {
+        handler.addMetaInfo(std::move(maybeMetaInfo.value()));
+      }
+    }
+  }
+
+  return handler;
+}
+
+std::vector<edm4hep::ParticleID> PIDHandler::getPIDs(const edm4hep::ReconstructedParticle& reco) const {
+  std::vector<edm4hep::ParticleID> pids;
+  const auto& [begin, end] = m_recoPidMap.equal_range(reco);
+  pids.reserve(std::distance(begin, end));
+
+  for (auto it = begin; it != end; ++it) {
+    pids.emplace_back(it->second);
+  }
+
+  return pids;
+}
+
+std::optional<edm4hep::ParticleID> PIDHandler::getPID(const edm4hep::ReconstructedParticle& reco,
+                                                      int32_t algoType) const {
+  for (const auto& pid : getPIDs(reco)) {
+    if (pid.getAlgorithmType() == algoType) {
+      return pid;
```

Comment:
```suggestion
  const auto& [begin, end] = m_recoPidMap.equal_range(reco);
  for (auto it = begin; it != end; ++it) {
    if (it->getAlgorithmType() == algoType) {
      return *it;
```

The vector of PIDs doesn't need to be created

Reply:
I was going for some code reuse, but good point. Changed.